File: autovisitor.py
#!/usr/bin/env python3
"""
AutoVisitor - Boost website traffic using Webshare rotating proxies
Usage: python autovisitor.py <url> [options]
"""

import os
import sys
import time
import random
import argparse

try:
    import requests
except ImportError:
    print("Error: 'requests' library not found. Run: pip install requests")
    sys.exit(1)

# ── Webshare API ──────────────────────────────────────────────────────────────
WEBSHARE_API_BASE = "https://proxy.webshare.io/api/v2"

# ── Random User-Agents to mimic real browsers ─────────────────────────────────
USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1",
    "Mozilla/5.0 (iPad; CPU OS 17_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4 Mobile/15E148 Safari/604.1",
    "Mozilla/5.0 (Android 14; Mobile; rv:125.0) Gecko/125.0 Firefox/125.0",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36 Edg/125.0.0.0",
]

ACCEPT_HEADERS = [
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
]


# ── Webshare Functions ────────────────────────────────────────────────────────

def get_proxy_list(token, mode="direct", page_size=100):
    """
    Fetch proxy list from Webshare API.
    mode: 'direct' or 'backbone'
    """
    headers = {"Authorization": "Token {}".format(token)}
    params = {"mode": mode, "page": 1, "page_size": page_size}
    try:
        resp = requests.get(
            "{}/proxy/list/".format(WEBSHARE_API_BASE),
            headers=headers,
            params=params,
            timeout=15,
        )
        resp.raise_for_status()
        data = resp.json()
        return data.get("results", [])
    except requests.exceptions.HTTPError as e:
        if e.response is not None and e.response.status_code == 401:
            print("[ERROR] Invalid Webshare API token. Check WEBSHARE_API_TOKEN.")
        else:
            print("[ERROR] Failed to fetch proxies: {}".format(e))
        return []
    except Exception as e:
        print("[ERROR] Could not connect to Webshare API: {}".format(e))
        return []


def format_proxy(proxy_obj):
    """Convert Webshare proxy object to requests-compatible dict."""
    username = proxy_obj.get("username", "")
    password = proxy_obj.get("password", "")
    host = proxy_obj.get("proxy_address", "")
    port = proxy_obj.get("port", 80)
    proxy_url = "http://{}:{}@{}:{}".format(username, password, host, port)
    return {"http": proxy_url, "https": proxy_url}


# ── Visitor Function ──────────────────────────────────────────────────────────

def visit_url(url, proxy_obj, timeout=15, verbose=False):
    """
    Send an HTTP GET request to url through the given proxy.
    Returns (success: bool, status_code: int or None, proxy_addr: str)
    """
    proxy_addr = "{}:{}".format(proxy_obj.get("proxy_address", ""), proxy_obj.get("port", ""))
    proxies = format_proxy(proxy_obj)
    headers = {
        "User-Agent": random.choice(USER_AGENTS),
        "Accept": random.choice(ACCEPT_HEADERS),
        "Accept-Language": "en-US,en;q=0.9",
        "Accept-Encoding": "gzip, deflate, br",
        "Connection": "keep-alive",
        "Cache-Control": "max-age=0",
        "Upgrade-Insecure-Requests": "1",
    }

    try:
        resp = requests.get(url, proxies=proxies, headers=headers, timeout=timeout, allow_redirects=True)
        if verbose:
            print("       Response: {} | {} bytes".format(resp.status_code, len(resp.content)))
        return (resp.status_code < 400, resp.status_code, proxy_addr)
    except requests.exceptions.ProxyError:
        if verbose:
            print("       Proxy error (skipping)")
        return (False, None, proxy_addr)
    except requests.exceptions.ConnectTimeout:
        if verbose:
            print("       Timeout (skipping)")
        return (False, None, proxy_addr)
    except Exception as e:
        if verbose:
            print("       Error: {}".format(e))
        return (False, None, proxy_addr)


# ── CLI ───────────────────────────────────────────────────────────────────────

def parse_args():
    parser = argparse.ArgumentParser(
        description="AutoVisitor — boost traffic to a URL using Webshare rotating proxies",
        formatter_class=argparse.RawTextHelpFormatter,
        epilog="""
Examples:
  python autovisitor.py https://myblog.com
  python autovisitor.py https://myblog.com -n 50 -d 2
  python autovisitor.py https://myblog.com -n 100 -d 0.5 --mode backbone -v
  python autovisitor.py https://myblog.com --token YOUR_TOKEN_HERE
        """,
    )
    parser.add_argument("url", help="Target URL to visit")
    parser.add_argument("-n", "--count", type=int, default=10,
                        help="Number of visits to send (default: 10)")
    parser.add_argument("-d", "--delay", type=float, default=1.0,
                        help="Delay in seconds between visits (default: 1.0)")
    parser.add_argument("--mode", choices=["direct", "backbone"], default="direct",
                        help="Webshare proxy mode: 'direct' or 'backbone' (default: direct)")
    parser.add_argument("--timeout", type=int, default=15,
                        help="Request timeout in seconds (default: 15)")
    parser.add_argument("--token", default="",
                        help="Webshare API token (overrides WEBSHARE_API_TOKEN env var)")
    parser.add_argument("-v", "--verbose", action="store_true",
                        help="Show detailed output per visit")
    return parser.parse_args()


def main():
    args = parse_args()

    # ── Resolve API token ──────────────────────────────────────────────────────
    token = args.token or os.environ.get("WEBSHARE_API_TOKEN", "")
    if not token:
        print("[ERROR] Webshare API token not found.")
        print("        Set env var WEBSHARE_API_TOKEN=<token>  or use --token <token>")
        sys.exit(1)

    print("=" * 60)
    print("  AutoVisitor")
    print("=" * 60)
    print("  Target  : {}".format(args.url))
    print("  Visits  : {}".format(args.count))
    print("  Delay   : {}s".format(args.delay))
    print("  Mode    : {}".format(args.mode))
    print("=" * 60)

    # ── Fetch proxies ──────────────────────────────────────────────────────────
    print("\n[*] Fetching proxy list from Webshare...")
    proxies = get_proxy_list(token, mode=args.mode)
    if not proxies:
        print("[ERROR] No proxies available. Check your Webshare plan or API token.")
        sys.exit(1)
    print("[+] Loaded {} proxies.\n".format(len(proxies)))

    # ── Visit loop ─────────────────────────────────────────────────────────────
    success_count = 0
    failed_count = 0

    for i in range(args.count):
        proxy = random.choice(proxies)
        proxy_addr = "{}:{}".format(proxy.get("proxy_address", ""), proxy.get("port", ""))

        print("[{}/{}] Visiting via {} ...".format(i + 1, args.count, proxy_addr), end="")
        sys.stdout.flush()

        ok, status, _ = visit_url(args.url, proxy, timeout=args.timeout, verbose=args.verbose)

        if ok:
            success_count += 1
            status_str = "OK ({})".format(status) if status else "OK"
            print("  [SUCCESS] {}".format(status_str))
        else:
            failed_count += 1
            status_str = "({})".format(status) if status else "(no response)"
            print("  [FAILED] {}".format(status_str))

        if i < args.count - 1:
            time.sleep(args.delay)

    # ── Summary ────────────────────────────────────────────────────────────────
    total = success_count + failed_count
    rate = (success_count / total * 100) if total > 0 else 0
    print("\n" + "=" * 60)
    print("  DONE")
    print("=" * 60)
    print("  Total   : {}".format(total))
    print("  Success : {} ({:.1f}%)".format(success_count, rate))
    print("  Failed  : {}".format(failed_count))
    print("=" * 60)


if __name__ == "__main__":
    main()

File: artifacts/api-server/src/routes/autovisitor.ts
import { Router, type Request, type Response } from "express";
import { randomUUID } from "crypto";
import { ProxyAgent, fetch as undiciFetch } from "undici";
import {
  StartSessionBody,
  GetSessionStatusParams,
  StopSessionParams,
} from "@workspace/api-zod";

const router = Router();

// ── User-Agent pool ───────────────────────────────────────────────────────────
const USER_AGENTS = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
  "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1",
  "Mozilla/5.0 (Android 14; Mobile; rv:125.0) Gecko/125.0 Firefox/125.0",
];

// ── Types ─────────────────────────────────────────────────────────────────────

type VisitLog = {
  index: number;
  proxy: string;
  success: boolean;
  statusCode: number | null;
  timestamp: string;
};

type SessionState = "pending" | "running" | "done" | "stopped" | "error";

type Session = {
  sessionId: string;
  state: SessionState;
  total: number;
  success: number;
  failed: number;
  completed: number;
  proxyCount: number | null;
  errorMessage: string | null;
  logs: VisitLog[];
  stopRequested: boolean;
};

// ── In-memory session store ───────────────────────────────────────────────────
const sessions = new Map<string, Session>();

// Clean up sessions older than 30 minutes
setInterval(() => {
  const cutoff = Date.now() - 30 * 60 * 1000;
  for (const [id, _] of sessions) {
    // Use sessionId creation time embedded in the id — just clear old done/stopped sessions
    const s = sessions.get(id);
    if (s && (s.state === "done" || s.state === "stopped" || s.state === "error")) {
      sessions.delete(id);
    }
  }
}, 5 * 60 * 1000);

// ── Webshare proxy fetcher ────────────────────────────────────────────────────
async function fetchProxies(token: string, mode: string): Promise<Array<{
  proxy_address: string;
  port: number;
  username: string;
  password: string;
}>> {
  const proxies: Array<{ proxy_address: string; port: number; username: string; password: string }> = [];
  let page = 1;

  while (true) {
    const url = `https://proxy.webshare.io/api/v2/proxy/list/?mode=${mode}&page=${page}&page_size=100`;
    const resp = await fetch(url, {
      headers: { Authorization: `Token ${token}` },
      signal: AbortSignal.timeout(15000),
    });

    if (!resp.ok) {
      if (resp.status === 401) throw new Error("Invalid Webshare API token");
      throw new Error(`Webshare API error: ${resp.status}`);
    }

    const data = (await resp.json()) as { results: typeof proxies; next: string | null };
    proxies.push(...data.results);
    if (!data.next) break;
    page++;
  }

  return proxies;
}

// ── Visit a URL through a proxy ───────────────────────────────────────────────
async function visitUrl(
  targetUrl: string,
  proxy: { proxy_address: string; port: number; username: string; password: string },
  timeoutMs = 15000,
): Promise<{ success: boolean; statusCode: number | null }> {
  const proxyUrl = `http://${proxy.username}:${proxy.password}@${proxy.proxy_address}:${proxy.port}`;
  const dispatcher = new ProxyAgent(proxyUrl);
  const ua = USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];

  try {
    const resp = await undiciFetch(targetUrl, {
      dispatcher,
      headers: {
        "User-Agent": ua,
        Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        "Cache-Control": "max-age=0",
        "Upgrade-Insecure-Requests": "1",
      },
      signal: AbortSignal.timeout(timeoutMs),
    });
    return { success: resp.status < 400, statusCode: resp.status };
  } catch {
    return { success: false, statusCode: null };
  }
}

// ── Background runner ─────────────────────────────────────────────────────────
async function runSession(
  session: Session,
  proxies: Array<{ proxy_address: string; port: number; username: string; password: string }>,
  targetUrl: string,
  delay: number,
) {
  session.state = "running";
  session.proxyCount = proxies.length;

  // Shuffle proxies for deterministic rotation
  const pool = [...proxies].sort(() => Math.random() - 0.5);
  const MAX_LOGS = 200;

  for (let i = 0; i < session.total; i++) {
    if (session.stopRequested) {
      session.state = "stopped";
      return;
    }

    const proxy = pool[i % pool.length];
    const proxyLabel = `${proxy.proxy_address}:${proxy.port}`;

    const { success, statusCode } = await visitUrl(targetUrl, proxy);

    if (success) session.success++;
    else session.failed++;
    session.completed++;

    const log: VisitLog = {
      index: i + 1,
      proxy: proxyLabel,
      success,
      statusCode,
      timestamp: new Date().toISOString(),
    };

    session.logs.unshift(log);
    if (session.logs.length > MAX_LOGS) session.logs.pop();

    if (i < session.total - 1 && delay > 0) {
      await new Promise<void>((resolve) => {
        let done = false;
        const timer = setTimeout(() => {
          done = true;
          clearInterval(check);
          resolve();
        }, delay * 1000);
        const check = setInterval(() => {
          if (session.stopRequested && !done) {
            done = true;
            clearTimeout(timer);
            clearInterval(check);
            resolve();
          }
        }, 100);
      });
    }
  }

  session.state = "done";
}

// ── Routes ────────────────────────────────────────────────────────────────────

// POST /api/autovisitor/start
router.post("/autovisitor/start", async (req: Request, res: Response) => {
  const parsed = StartSessionBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const { token, url, count, delay = 1.0, mode = "direct" } = parsed.data;

  // Basic URL validation
  try {
    new URL(url);
  } catch {
    res.status(400).json({ error: "Invalid URL format" });
    return;
  }

  const sessionId = randomUUID();
  const session: Session = {
    sessionId,
    state: "pending",
    total: count,
    success: 0,
    failed: 0,
    completed: 0,
    proxyCount: null,
    errorMessage: null,
    logs: [],
    stopRequested: false,
  };
  sessions.set(sessionId, session);

  // Fetch proxies and start visits in background
  (async () => {
    try {
      const proxies = await fetchProxies(token, mode);
      if (proxies.length === 0) {
        session.state = "error";
        session.errorMessage = "No proxies available. Check your Webshare plan.";
        return;
      }
      await runSession(session, proxies, url, delay);
    } catch (err) {
      session.state = "error";
      session.errorMessage = err instanceof Error ? err.message : "Unknown error";
    }
  })();

  res.json({ sessionId, status: "started" });
});

// GET /api/autovisitor/status/:sessionId
router.get("/autovisitor/status/:sessionId", (req: Request, res: Response) => {
  const parsed = GetSessionStatusParams.safeParse(req.params);
  if (!parsed.success) {
    res.status(400).json({ error: "Invalid session ID" });
    return;
  }

  const session = sessions.get(parsed.data.sessionId);
  if (!session) {
    res.status(404).json({ error: "Session not found" });
    return;
  }

  const { stopRequested: _, ...statusData } = session;
  res.json(statusData);
});

// POST /api/autovisitor/stop/:sessionId
router.post("/autovisitor/stop/:sessionId", (req: Request, res: Response) => {
  const parsed = StopSessionParams.safeParse(req.params);
  if (!parsed.success) {
    res.status(400).json({ error: "Invalid session ID" });
    return;
  }

  const session = sessions.get(parsed.data.sessionId);
  if (!session) {
    res.status(404).json({ error: "Session not found" });
    return;
  }

  session.stopRequested = true;
  const { stopRequested: _, ...statusData } = session;
  res.json(statusData);
});

export default router;
